### ISSUE TITLE
Switch & case

### ISSUE THREAD

**Issue by nfischer** [nfischer/shelljs-transpiler#16]
Add support for switch & case, including the `default` case (notated by `*)`).

This should also add support for things like:

```bash
?)
  echo "do something"
  ;;
```

Also, this should handle the `;&` fall-through syntax (although that's less of a priority, since it's only for newer versions of bash).

**Comment by Schmavery** [comment]
Hey @nfischer
I was playing with this a little on my end and I've got something like this right now:
![image](https://cloud.githubusercontent.com/assets/2154522/18732074/389ff2f8-8015-11e6-95ff-009669193207.png)

This is almost enough for my purposes except that I noticed that it's possible to do something like 
```bash
#!/bin/sh

FRUIT="kiwi"

case "$FRUIT" in
   a*) echo "Fruit starts with the letter a." 
   ;;
   b*) echo "Fruit starts with the letter a." 
   ;;
esac
```

I'm sure other similar patterns are also possible.  I guess given this, we'd have to use if/else if instead of the javascript switch statement.  Any idea of a way around this other than writing code to do the pattern matching against ` * `?  On the upside, if we can solve this problem, I'd imagine we don't have to special case the default ` *) ` case.  I'm no bash expert so I'd imagine there are a lot of other edge cases in the bash case grammar that I'm missing (I guess the ` ? ` in your example is just another wildcard pattern?).
Are these patterns only used in case statements?  If not, maybe we can figure out a way to solve this everywhere.

**Comment by nfischer** [comment]
Bash supports glob patterns in its case statements. Any wildcard you could type on your shell, is also supported as a case label from what I've seen. You're correct, this definitely complicates the implementation.

Globs are not equivalent to regex, but they're easily translatable to regex. Perhaps we should use an if-else ladder with regex tests.

To keep things looking pretty, we could also default back to a standard switch statement for non-globby case patterns.

**Comment by Schmavery** [comment]
Any idea if shelljs or equivalent has any such translator from glob to regex?

**Comment by nfischer** [comment]
I'm a maintainer over at shelljs too. We don't support a function like what you described, but the post you linked to is precisely along the lines of what I was thinking 👍🏻

**Comment by Schmavery** [comment]
In case you're interested, I've got it working like this now:
![image](https://cloud.githubusercontent.com/assets/2154522/18735608/f411cf5c-8033-11e6-948c-6b642acc6d37.png)

**Comment by nfischer** [comment]
@Schmavery that looks awesome! I'd be happy to take a look at a pull request :smile:

**Comment by Schmavery** [comment]
I'm trying to clean this up for a PR, but I'm having trouble figuring out how ohm/the transpiler solves a specific kind of problem:
In my specification of the grammar for this construct, I have a rule (SwitchCommand) for the outer `case expr in esac` bit, and then SwitchCommand contains a list of "CaseCommand"s (the inner cases).  In my construction for this, I need a variable that is defined in the parent SwitchCommand rule but that the inner CaseCommand rules have access to.  Any idea how we can pass context to child elements when generating code?

**Comment by nfischer** [comment]
@Schmavery I think this sort of thing is somewhat difficult to do in Ohm. It's crept up a couple times for me, and I've only made hacks around it, so feel free to post the question on the mailing list (ohm@cdglabs.org) :smile:

In my case, I needed to set some variable in the outer rule, access it in the inner rules, and then unset it once I'm done with the whole outer rule. It sounded to me a lot like pushing and popping from a stack, so I snuck by using a global variable on the function call stack:

```js
var value = 'some initial value';
function outer(inner) {
  var oldVal = value; // save the current value in a local variable
  value = 'foo'; // assign some new value
  inner.toJS(); // process each of the inner rules, which can read from the global value
  value = oldVal; // restore the old value to handle nesting
}
```

As far as I can tell, this should adequately handle nesting switch statements as well (although I can't recall if I've vigorously tested this). Also, feel free to open a PR if you're stuck and I can chime in where I can.
